### DISCUSSION TITLE
Integration of own Vulkan renderer

### DISCUSSION THREAD

**Discussion by llfab** [AvaloniaUI/Avalonia#7137] [Q&A] [answered]
In a previous discussion (https://github.com/AvaloniaUI/Avalonia/discussions/6842) we had addressed the technical aspects of integrating a custom OpenGL renderer. After some back and forth, we have decided to not implement a new OpenGL backend to our 3D renderer (which already has a DX 11 backend) but implement the new cross-platform backend in Vulkan. In order to understand how to integrate this in Avalonia with as little overhead as possible, I am opening this separate thread. 
To understand better here some requirments / setup aspects of our 3D renderer:
- We typically display not so complex 3D scenes (compared to gaming) but require simple geometry, overlay, mesh and especially volume rendering
- The 3D scene views are embedded as "controls" into the 2D UI. We use many scenes and scene views within the same UI application
- The 3D scene view shall be overlayable with controls from the surrounding 2D UI framwork. Therefore, the rendered texture from our 3 renderer must be "nicely" integrated into the 2D framework.
- We must support Windows and Linux. MacOS is currently optional.

Currently there seem these options:
1. Take the Vulkan texture and put its bitmap into something like a writeable bitmap control
2. Use Vulkan / OpenGL texture sharing (https://stackoverflow.com/questions/38907764/how-to-render-to-opengl-from-vulkan)
3. [EDIT] Use Vulkan rendering backend for Avalonia/Skia and just wrap the Vulkan texture as SKImage and "paint" that in a custom control

(1.) has the advantage that it is very simple to implement but the drawback that the texture(s) are downloaded from the GPU memory, taken into the writeable bitmap and finally uploaded (again) to the GPU memory.

(2.) is a more complex approach but should provide optimal efficiency.

(3.) [EDIT] should provide optimal efficiency, no complex OpenGL / Vulkan interaction, unclear how "stable" Avalonia is with Vulkan backend.

Hence, we are working on making the Sample from Khronos group for Vulkan/OpenGL texture sharing work for Avalonia. This seems to NOT work for the standard Avalonia package 0.10.10. Therefore, we are basing this now directly on the Avalonia source code using the opentk-integration branch. 

I will let you know about the progress. And there will be more questions from my side - I believe...

**Comment by kekekeks** [comment, marked as answer]
Regarding Avalonia itself.
Windows: we render through ANGLE->DirectX11 (there is a DX9 fallback for old Windows versions), so DXGI texture sharing is possible (that's how WGL-produced frames are drawn with the current OpenTK PR) but not yet exposed as an API.
Linux: we usually use GLX, but there is an option for EGL. Technically EGL allows passing libgbm-allocated DMA-BUF handles, but extension is relatively new and not yet well supported by NVIDIA. We use OpenGL context sharing for user OpenGL rendering.
macOS: we use CGL to create OpenGL contexts and render into an IOSurface (macOS DXGI handle counterpart) which is later assigned to CALayer. We use OpenGL context sharing for user OpenGL rendering.


The low-level APIs with more control are:
- `IPlatformOpenGlInterface` - allows creation of OpenGL contexts that either use OpenGL context sharing (Linux/macOS) or are compatible with OS texture sharing (currently DXGI, will add IOSurface later).
- `OpenGlBitmap` - can be used as a regular bitmap, but also has `CreateFramebufferAttachment` method which attaches a render target  to the current FBO's `GL_COLOR_ATTACHMENT0`. Returned object also has `Present()` method that does the necessary steps to transfer the render results to Avalonia's render thread.

**Reply by MarchingCube** [reply]
You can also configure Avalonia to render using Vulkan (that needs some setup to get right) and wrap your renderer textures as Skia images to render directly from Skia. This avoids copying and is very fast. We also tried to use OpenGL before we moved to Vulkan but it was not worth the effort (Avalonia renders on a separate thread by default, one can make it kinda work with context sharing but it is fairly slow due to context switching cost)

This is completely doable even in current stock Avalonia and SkiaSharp (I'm running Avalonia on Vulkan from two years).

**Reply by llfab** [reply]
This sounds like a nice and less complex option compared to texture sharing. So:
- I did not find docu on how to run Avalonia with the Skia Vulkan backend. Anything you can guide us to?
- Do you know of any sample implementation for passing the rendered Vulkan texture directly to Skia via SkImage or similar?
- Any known limitations to this approach?

**Reply by MarchingCube** [reply]
You need to implement a custom `ISkiaGpu` (you can take a look at `GlSkiaGpu`) and create Skia `GRContext` that is Vulkan based. There are examples in SkiaSharp repo too from what I remember.

For creating `SKImage` from vulkan textures - take a look at `SKImage.FromTexture` which allows for wrapping backend textures.

So far I don't see any big limitations with this approach, it takes a while to get right due to Vulkan synchronization (you need to deal with fences and barriers to make sure that your textures are ready since Skia does no layout transitions).

**Reply by llfab** [reply]
I have looked at what has been done here: https://github.com/mono/SkiaSharp/issues/252 and some of the code of this PR.

My feeling is, that this needs quite some effort to get a VkSkiaGpu up and running. In addition, it may need some continuous work while Avalonia is changing. It does seem though like something we could hope for Avalonia picking this up providing this implementation as part of Avalonia.Skia.

**Reply by llfab** [reply]
> You need to implement a custom `ISkiaGpu` (you can take a look at `GlSkiaGpu`) and create Skia `GRContext` that is Vulkan based. There are examples in SkiaSharp repo too from what I remember.
> 
> For creating `SKImage` from vulkan textures - take a look at `SKImage.FromTexture` which allows for wrapping backend textures.
> 
> So far I don't see any big limitations with this approach, it takes a while to get right due to Vulkan synchronization (you need to deal with fences and barriers to make sure that your textures are ready since Skia does no layout transitions).

If you have a working `VkSkiaGpu` implementation that includes the Vk infrastructure setup, maybe that's worth publishing. Could this not even be a push for Avalonia providing a Vk backend in the future...?

**Reply by MarchingCube** [reply]
I cannot publish my current implementation directly since it is a part of a product I'm working on (closed source). There is also a planned renderer work (composition) that would again change synchronization so that would need to happen first.

**Reply by llfab** [reply]
@MarchingCube : I saw that you were involved a bit in https://github.com/AvaloniaUI/Avalonia/pull/7753. Do you think this is on a solid basis an will be part of version [0.11](https://github.com/AvaloniaUI/Avalonia/projects/10) ?

**Comment by llfab** [comment]
@kekekeks : I have also looked at https://github.com/AvaloniaUI/Avalonia/issues/2176 which I believe is now provided as `ICustomDrawOperation` like in https://github.com/AvaloniaUI/Avalonia/blob/3c22ce5b1b9bd1dd66f9021696427992c11779b2/samples/RenderDemo/Pages/CustomSkiaPage.cs#L21.

I have played a bit with `WriteableBitmap `using uncompressed video frames to check the performance of it. I have used `<Image>` with its source bound to a `WriteableBitmap`. This works fine but seems to drop in function/performance significantly once the frame rate reaches more than 30 fps. Do you think custom WritableBitmap drawing in a custom skia page, would improve that? If that was the case, we may be able to do the Vulkan integration without (complex) shared context and without activating the Vulkan backend for Avalonia. Or do you guys have the feeling that using the Vulkan texture as a bitmap in a custom control would not perform nicely anyhow?

**Comment by llfab** [comment]
We have made Vulkan rendering work within an Avalonia control both on Linux and Windows. I may add some code samples at a later stage. For Windows we found that the integration works with Avalonia package only for WGL, which is not the best solution. In order to make it work with (default) EGL we needed to use the opentk-branch package because only that allowed for context / texture sharing.

Note that such an integration requires both vulkan and opengl extension functions wich may not be available on all systems and it certainly does not work on virtual machines. 

We will continue on this once we have version [0.11](https://github.com/AvaloniaUI/Avalonia/projects/10) available which should allow for such things.

Also see the related discussion: https://github.com/AvaloniaUI/Avalonia/discussions/8556#discussioncomment-3192546
